### PR TITLE
Add some render box updates

### PR DESCRIPTION
**Pull request by renancaraujo**
# Description

Some improvements on our render object widget. 

- Made game loop call dispose on the ticker on detach. 
- Made the render object a repaint boundary, avoid an entire screen that contains a GameWidget to repaint every frame.
- Made the render object itself as expanded enabling the removal of that `RenderConstrainedBox`

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors and are passing (See [Contributor Guide]).
- [x] My PR does not decrease the code coverage, or I have __a very special case__ and explained on the PR description why this PR decreases the coverage.
- [x] I updated/added relevant documentation (doc comments with `///`) and updated/added examples in `doc/examples`.
- [x] I have formatted my code with `./scripts/format.sh` and the Flame analyzer (`./scripts/analyze.sh`) does not report any problems.
- [x] I read and followed the [Flame Style Guide].
- [x] I have added a description of the change under `[next]` in `CHANGELOG.md`.
- [x] I removed the `Draft` status, by clicking on the `Ready for review` button in this PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require Flame users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in `CHANGELOG.md`).
- [x] No, this is *not* a breaking change.

## Related Issues

[issue database]: https://github.com/flame-engine/flame/issues
[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Flame Style Guide]: https://github.com/flame-engine/flame/blob/main/STYLEGUIDE.md

### CODE REVIEW

File: packages/flame/CHANGELOG.md [hunk 1]
```diff
@@ -23,6 +23,7 @@
  - Add `SpriteAnimation` support to parallax
  - Fix `Parallax` alignment for images with different width and height
  - Fix `ImageComposition` image bounds validation
+ - Improved the internal Render object widget performance
```
Comment (wolfenrain) [inline packages/flame/CHANGELOG.md:26]:
```suggestion
 - Improved the internal `RenderObject` widget performance
```

**Review by wolfenrain — APPROVED** [review]
LGTM

**Review by japalekhin — APPROVED** [review]

**Comment by erickzanardo** [comment]
@renancaraujo does this fixes #834 ?

**Comment by renancaraujo** [comment]
> @renancaraujo does this fixes #834 ?

No

**Review by erickzanardo — APPROVED** [review]

File: packages/flame/lib/src/game/game_widget/game_widget.dart [hunk 2]
```diff
@@ -293,9 +293,11 @@ class _GameRenderObjectWidget extends LeafRenderObjectWidget {
...
-    return RenderConstrainedBox(
-      child: GameRenderBox(context, game),
-      additionalConstraints: const BoxConstraints.expand(),
-    );
+    return GameRenderBox(context, game);
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, GameRenderBox renderObject) {
+    renderObject..game = game;
```
Comment (spydon) [inline packages/flame/lib/src/game/game_widget/game_widget.dart:301]: What is it that happens here?

Comment (erickzanardo-dx) [inline packages/flame/lib/src/game/game_widget/game_widget.dart:301]:
```suggestion
    renderObject.game = game;
```

File: packages/flame/lib/src/game/game_render_box.dart [hunk 3]
```diff
@@ -60,6 +63,11 @@ class GameRenderBox extends RenderBox with WidgetsBindingObserver {
     markNeedsPaint();
   }
 
+  @override
+  void performLayout() {
+    size = constraints.biggest;
```
Comment (luanpotter) [inline packages/flame/lib/src/game/game_render_box.dart:68]: how was this not a problem before? was it a bug?

**Review by luanpotter — APPROVED** [review]
LGTM as long as our examples that create & delete games at will work (ie no disposed resource is then accessed later)

File: packages/flame/lib/src/game/game_render_box.dart (hunk 3 above)
Comment (renancaraujo) [reply packages/flame/lib/src/game/game_render_box.dart:68]: We have compute dry.

**Review by spydon — APPROVED** [review]

File: examples/.gitignore [hunk 4]
```diff
@@ -1,3 +1,6 @@
+macos/
+test/
+
```
Comment (renancaraujo) [inline examples/.gitignore:3]: This change removes examples from having macos and test committed to the repo

File: packages/flame/lib/src/game/game_loop.dart [hunk 5]
```diff
@@ -28,6 +28,10 @@ class GameLoop {
     _ticker.stop();
   }
 
+  void dispose() {
+    _ticker.dispose();
+  }
```
Comment (renancaraujo) [inline packages/flame/lib/src/game/game_loop.dart:33]: Added access to dispose via our gameloop. The actual implementation of both tickers stop and dispose are quite similar in terms of freeing resources, but this is a matter of internal flutter implementation, we should use the public API as intended since internal stuff can change there.

File: packages/flame/lib/src/game/game_render_box.dart [hunk 6]
```diff
@@ -23,6 +23,9 @@ class GameRenderBox extends RenderBox with WidgetsBindingObserver {
   @override
   bool get sizedByParent => true;
 
+  @override
+  bool get isRepaintBoundary => true;
```
Comment (renancaraujo) [inline packages/flame/lib/src/game/game_render_box.dart:27]: This makes the "markNeedsPaint" not to spread to the parent widget, making out 60 fps repaint only affect things inside flame, and not the whole application.

File: packages/flame/lib/src/game/game_render_box.dart [hunk 7]
```diff
@@ -48,7 +51,7 @@ class GameRenderBox extends RenderBox with WidgetsBindingObserver {
   void detach() {
     super.detach();
     game.detach();
-    gameLoop.stop();
+    gameLoop.dispose();
```
Comment (renancaraujo) [inline packages/flame/lib/src/game/game_render_box.dart:54]: Call dispose on detaching to free sources.

File: packages/flame/lib/src/game/game_render_box.dart (hunk 3 above)
Comment (renancaraujo) [reply packages/flame/lib/src/game/game_render_box.dart:68]: That s called because we say to the framework that we are sized by the parent.

File: packages/flame/lib/src/game/game_widget/game_widget.dart (hunk 2 above)
Comment (renancaraujo) [reply packages/flame/lib/src/game/game_widget/game_widget.dart]: Render objects are log living, while widgets live and die every rebuild. Since parameters of a widget can change between builds, we may need to update the render object with the current instance of game. As of today, we don't react to that change.

**Review by renancaraujo — COMMENTED** [review]
> LGTM as long as our examples that create & delete games at will work (ie no disposed resource is then accessed later)

Made a change to make a gameloop live as long as an attachment. A render object shall never be reattached but that will work anyways.

File: packages/flame/lib/src/game/game_render_box.dart [hunk 8]
```diff
@@ -34,6 +37,8 @@ class GameRenderBox extends RenderBox with WidgetsBindingObserver {
     super.attach(owner);
     game.attach(owner, this);
 
+    final gameLoop = this.gameLoop = GameLoop(gameLoopCallback);
```
Comment (spydon) [inline packages/flame/lib/src/game/game_render_box.dart:40]: No need for this variable right? You can just set `this.gameLoop` and that will be in the scope for this method?
```suggestion
    this.gameLoop = GameLoop(gameLoopCallback);
```

File: packages/flame/lib/src/game/game_render_box.dart [hunk 9]
```diff
@@ -13,16 +13,19 @@ import 'game_loop.dart';
 class GameRenderBox extends RenderBox with WidgetsBindingObserver {
   BuildContext buildContext;
   Game game;
-  late GameLoop gameLoop;
+  GameLoop? gameLoop;
```
Comment (spydon) [inline packages/flame/lib/src/game/game_render_box.dart:16]: Can't this still be late?

Comment (renancaraujo) [reply packages/flame/lib/src/game/game_render_box.dart:16]: We should set null on detach.

File: packages/flame/lib/src/game/game_render_box.dart (hunk 8 above)
Comment (renancaraujo) [reply packages/flame/lib/src/game/game_render_box.dart:40]: Analyzer is dumb, not recnizing the non nullability there.

File: packages/flame/lib/src/game/game_render_box.dart (hunk 9 above)
Comment (spydon) [reply packages/flame/lib/src/game/game_render_box.dart:16]: Aah, makes sense

File: packages/flame/lib/src/game/game_render_box.dart (hunk 8 above)
Comment (spydon) [reply packages/flame/lib/src/game/game_render_box.dart:40]: Yeah, this was with the other comment in mind.
But then I guess this can be merged if you just run for formatter on it then? :)
